File: 06-linked_list/00023-Merge_k_Sorted_Lists/00023-merge_k_sorted_lists.go
// 23: Merge k Sorted Lists
// https://leetcode.com/problems/merge-k-sorted-lists

package main

import (
	"container/heap"
	"fmt"
)

type ListNode struct {
    val int
    next *ListNode
}

type LinkedList struct {
    head *ListNode
}

func llInput(head **ListNode, inputs []int) {
    for i:=len(inputs)-1; i>=0; i-- {
        node := &ListNode{inputs[i], *head}
        *head = node
    }
}

func llOutput(head *ListNode) {
    for head != nil {
        fmt.Print(head.val, " -> ")
        head = head.next
    }
    fmt.Println("NULL")
}

type minHeap []*ListNode
func (h minHeap) Len() int { return len(h) }
func (h minHeap) Less(i, j int) bool { return h[i].val < h[j].val }
func (h minHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
func (h *minHeap) Push(x interface{}) { *h = append(*h, x.(*ListNode)) }
func (h *minHeap) Pop() interface{} {
    min := (*h)[len(*h) - 1]
    *h = (*h)[:len(*h) - 1]
    return min
}

// SOLUTION
func mergeKLists(lists []*ListNode) *ListNode {
    pq := &minHeap{}
	heap.Init(pq)
	var head, tail, next *ListNode

	for i:=0; i<len(lists); i++ {
		if lists[i]!=nil {
			heap.Push(pq, lists[i])
		}
	}

	for pq.Len() > 0 {
		if tail != nil {
			next = heap.Pop(pq).(*ListNode)
			if next.next != nil {
				heap.Push(pq, next.next)
			}
			tail.next = next
			tail = next
		} else {
			head = heap.Pop(pq).(*ListNode)
			tail = head
			if head.next != nil {
				heap.Push(pq, head.next)
			}
		}
	}

	return head
}


func main() {
    ll1, ll2, ll3 := LinkedList{}, LinkedList{}, LinkedList{}
    
    // INPUT
    li1, li2, li3 := []int{1,4,5}, []int{1,3,4}, []int{2,6}
    llInput(&ll1.head, li1)
    llInput(&ll2.head, li2)
    llInput(&ll3.head, li3)
    lists := []*ListNode{ll1.head, ll2.head, ll3.head}

    // OUTPUT
    result := mergeKLists(lists)
    llOutput(result)
}

File: 06-linked_list/00023-Merge_k_Sorted_Lists/00023-merge_k_sorted_lists.py
# 23: Merge k Sorted Lists
# https://leetcode.com/problems/merge-k-sorted-lists


from queue import PriorityQueue


class ListNode:
    def __init__(self, x: int=None):
        self.val = x
        self.next = None

class LinkedList:
    def __init__(self) -> None:
        self.head = None

    def Input(self, inputs: list[int]):
        for i in inputs:
            node = ListNode(i)
            if self.head:
                tail = self.head
                while tail.next != None: tail = tail.next
                tail.next = node
            else:
                self.head = node

    def Output(self, head: ListNode):
        while (head != None):
            print(head.val, "-> ", end='')
            head = head.next
        print("NULL")


class Solution:
    # SOLUTION
    def mergeKLists(self, lists: list[ListNode]) -> ListNode:
        heap = PriorityQueue()
        head = ListNode(None)
        c = head

        for i, node in enumerate(lists):
            if node:
                heap.put((node.val, i, node))

        while heap.qsize()>0:
            t = heap.get()
            c.next, i = t[2], t[1]
            c = c.next
            if c.next:
                heap.put((c.next.val, i, c.next))
        
        return head.next


if __name__ == "__main__":
    o = Solution()
    ll1, ll2, ll3 = LinkedList(), LinkedList(), LinkedList()

    # INPUT
    li1, li2, li3 = [1,4,5], [1,3,4], [2,6]
    ll1.Input(li1)
    ll2.Input(li2)
    ll3.Input(li3)
    lists: list[ListNode] = [ll1.head, ll2.head, ll3.head]

    # OUTPUT
    result = o.mergeKLists(lists)
    ll1.Output(result)
